### DISCUSSION TITLE
[Feature Proposal] Provide a method to increase CSS specificity to ease development of component systems

### DISCUSSION THREAD

**Discussion by AndyOGo** [tailwindlabs/tailwindcss#10924] [Ideas]
I would like to propose an alternative method to increase [CSS specificity](https://developer.mozilla.org/en-US/docs/Web/CSS/Specificity) to the [`!` important modifier](https://tailwindcss.com/docs/configuration#important-modifier).

I suggest to support an **adjoining classes** modifier like `trump:` or `bump:`.

It would render the utility like:

```html
<div class="red bump:blue">
```

**Generated CSS:**
```css
.blue,
.bump.blue { // <- this adjoining class wins
  color: blue;
}

red {
  color: red;
}
```

A [simple live example](https://html-css-js.com/?html=%3C!DOCTYPE%20html%3E%0A%3Chtml%3E%0A%3Chead%3E%0A%3Ctitle%3EHTML%20CSS%20JS%3C/title%3E%0A%3C/head%3E%0A%3Cbody%3E%0A%20%20%3Cdiv%20class=%22blue%22%3E%3C/div%3E%0A%20%20%3Cdiv%20class=%22red%22%3E%3C/div%3E%0A%0A%20%20%3Cdiv%20class=%22blue%20red%22%3E%3C/div%3E%0A%20%20%3Cdiv%20class=%22blue%20trump%20red%22%3E%3C/div%3E%0A%3C/body%3E%0A%3C/html%3E&css=/*%20CSS%20styles%20*/%0A.blue,%0A.red%20%7B%0A%20%20wi$*$dth:%20100px;%0A%20%20height:%2020px;%0A%20%20margin-top:%2010px;%0A%7D%0A%0A.blue,%0A.blue.trump%20%7B%0A%20%20background%20:%20blue;%0A%7D%0A%0A.red%20%7B%0A%20%20background:%20red;%0A%7D%0A%09%20%20&js=%0A).

---

**Background:**

Using tailwind for component or design systems often has the need to override tailwind classes of a component, e.g. colors or dimensions.

I found these posts explain the problem at hand very well.
- https://dev.to/diogoko/overriding-tailwind-classes-in-react-4po3
- https://sancho.dev/blog/tailwind-and-design-systems

Related issues:
- https://github.com/tailwindlabs/tailwindcss/issues/9980
- https://github.com/tailwindlabs/tailwindcss/issues/1013
- https://github.com/tailwindlabs/tailwindcss/issues/1432
- https://github.com/tailwindlabs/tailwindcss/issues/436

**Comment by olets** [comment]
This is a good use case for `addVariant()`. You can add your own `bump:` variant. https://tailwindcss.com/docs/plugins#static-variants

Fwiw the solution I'm used to seeing for the specific problem you described is to add a prop for "color"/"intent"/etc, and then have logic to say something like "color with a fallback of red" or "if intent is primary, red. if intent is secondary, blue". In some cases, that can have added self-documentation benefits.

**Comment by AndyOGo** [comment]
I published [`tailwindcss-trump`](https://www.npmjs.com/package/tailwindcss-trump).

The static variant is super easy and simply repeats the selector.

> **Note:** Repeated occurrences of the same simple selector are allowed and do increase specificity.
https://www.w3.org/TR/selectors-3/#specificity

```js
const plugin = require('tailwindcss/plugin')

module.exports = {
  // ...
  plugins: [
    plugin(function({ addVariant }) {
      addVariant('trump', '&&')
    })
  ]
}
```

Play with it here
https://play.tailwindcss.com/v5ZpuKy2hH

**Reply by remus-selea** [reply]
Could this be migrated to v4?

**Reply by AndyOGo** [reply]
@remus-selea It still works with tailwindcss 4 `@plugin` directive for legacy JS plugins
https://tailwindcss.com/docs/functions-and-directives#plugin-directive

In your config:
> @plugin "tailwindcss-trump";

Example
https://play.tailwindcss.com/jwYgxSVsZI?file=css

A part from that peer dependencies should be updated
https://github.com/AndyOGo/tailwindcss-trump/blob/ed87fb54c75db2876817358705877abb1850b63e/package.json#L57
